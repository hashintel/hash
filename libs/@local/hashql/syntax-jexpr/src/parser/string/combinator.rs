use hashql_ast::heap::{self, Heap};
use winnow::{
    Parser,
    ascii::multispace0,
    combinator::{delimited, separated_foldl1, trace},
    error::ParserError,
    stream::{AsChar, Stream, StreamIsPartial},
};

enum VecOrOneValue<'heap, T> {
    Vec(heap::Vec<'heap, T>),
    One(T),
}

impl<'heap, T> VecOrOneValue<'heap, T> {
    fn as_vec<'this>(
        this: &'this mut Option<Self>,
        heap: &'heap Heap,
    ) -> &'this mut heap::Vec<'heap, T> {
        // Vec won't allocate until it is pushed to
        let value = this.take().unwrap_or_else(|| Self::Vec(heap.vec(None)));

        let value = match value {
            Self::Vec(value) => value,
            Self::One(value) => {
                let mut vec = heap.vec(None);
                vec.push(value);
                vec
            }
        };

        let value = this.insert(Self::Vec(value));
        match value {
            Self::Vec(value) => value,
            Self::One(_) => unreachable!(),
        }
    }
}

struct VecOrOne<'heap, T> {
    heap: &'heap Heap,
    value: Option<VecOrOneValue<'heap, T>>,
}

impl<'heap, T> VecOrOne<'heap, T> {
    const fn new(heap: &'heap Heap, value: T) -> Self {
        Self {
            heap,
            value: Some(VecOrOneValue::One(value)),
        }
    }

    fn into_vec(self) -> heap::Vec<'heap, T> {
        match self.value {
            Some(VecOrOneValue::Vec(vec)) => vec,
            Some(VecOrOneValue::One(value)) => {
                let mut vec = self.heap.vec(None);
                vec.push(value);
                vec
            }
            None => self.heap.vec(None),
        }
    }

    pub(crate) fn append(&mut self, value: &mut Self) -> &mut Self {
        let Some(value) = value.value.take() else {
            return self;
        };

        let vec = VecOrOneValue::as_vec(&mut self.value, self.heap);
        match value {
            VecOrOneValue::Vec(mut values) => {
                vec.append(&mut values);
            }
            VecOrOneValue::One(value) => {
                vec.push(value);
            }
        }

        self
    }
}

pub(crate) fn separated_alloc1<Input, Output, Sep, Error, ParseNext, SepParser>(
    heap: &Heap,
    parser: ParseNext,
    sep: SepParser,
) -> impl Parser<Input, heap::Vec<'_, Output>, Error>
where
    Input: Stream,
    ParseNext: Parser<Input, Output, Error>,
    SepParser: Parser<Input, Sep, Error>,
    Error: ParserError<Input>,
{
    trace(
        "separated_alloc1",
        separated_foldl1(
            parser.map(|value| VecOrOne::new(heap, value)),
            sep,
            |mut left, _, mut right| {
                left.append(&mut right);
                left
            },
        )
        .map(VecOrOne::into_vec),
    )
}

/// A combinator that takes a parser `inner` and produces a parser that also consumes both
/// leading and trailing whitespace, returning the output of `inner`.
pub(crate) fn ws<Input, Output, Error, ParseNext>(
    parser: ParseNext,
) -> impl Parser<Input, Output, Error>
where
    Input: StreamIsPartial + Stream<Token: AsChar + Clone>,
    ParseNext: Parser<Input, Output, Error>,
    Error: ParserError<Input>,
{
    trace(
        "ws", //
        delimited(multispace0, parser, multispace0),
    )
}
