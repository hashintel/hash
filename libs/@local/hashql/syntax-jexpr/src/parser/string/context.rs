use core::ops::Range;

use hashql_ast::heap::Heap;
use hashql_core::span::{SpanId, TextRange, TextSize, storage::SpanStorage};
use winnow::{LocatingSlice, Stateful};

use crate::span::Span;

#[expect(
    clippy::cast_possible_truncation,
    reason = "The lexer ensures that we never have more than 4GiB of text"
)]
fn range_to_text(range: Range<usize>) -> TextRange {
    TextRange::new(
        TextSize::from((range.start + 1) as u32),
        TextSize::from((range.end + 1) as u32),
    )
}

#[derive(Debug, Copy, Clone)]
pub(crate) struct Context<'heap, 'span> {
    pub heap: &'heap Heap,
    pub spans: &'span SpanStorage<Span>,
    pub parent: SpanId,
}

impl Context<'_, '_> {
    pub(crate) fn span(&self, range: Range<usize>) -> SpanId {
        // `+ 1` here to offset the opening quote
        self.spans.insert(Span {
            range: range_to_text(range),
            pointer: None,
            parent_id: Some(self.parent),
        })
    }

    pub(crate) fn cover(&self, span: SpanId, range: Range<usize>) {
        self.spans.modify(span, |span| {
            span.range = span.range.cover(range_to_text(range.clone()));
        });
    }
}

pub(crate) type Input<'heap, 'span, 'source> =
    Stateful<LocatingSlice<&'source str>, Context<'heap, 'span>>;
