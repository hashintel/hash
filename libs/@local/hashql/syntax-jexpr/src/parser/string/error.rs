use alloc::borrow::Cow;
use core::fmt::Write as _;

use hashql_core::span::{SpanId, storage::SpanStorage};
use hashql_diagnostics::{
    Diagnostic,
    category::{DiagnosticCategory, TerminalDiagnosticCategory},
    help::Help,
    label::Label,
    note::Note,
    severity::Severity,
};
use text_size::{TextRange, TextSize};
use winnow::error::{ContextError, ParseError, StrContext};

use crate::span::Span;

pub(crate) type StringDiagnostic = Diagnostic<StringDiagnosticCategory, SpanId>;

const INVALID_EXPR: TerminalDiagnosticCategory = TerminalDiagnosticCategory {
    id: "invalid-expression",
    name: "Invalid string expression",
};

#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
pub enum StringDiagnosticCategory {
    InvalidExpression,
}

impl DiagnosticCategory for StringDiagnosticCategory {
    fn id(&self) -> Cow<'_, str> {
        Cow::Borrowed("string")
    }

    fn name(&self) -> Cow<'_, str> {
        Cow::Borrowed("String")
    }

    fn subcategory(&self) -> Option<&dyn DiagnosticCategory> {
        match *self {
            Self::InvalidExpression => Some(&INVALID_EXPR),
        }
    }
}

#[expect(
    clippy::cast_possible_truncation,
    reason = "lexer ensures we never parse more than 4GiB"
)]
pub(crate) fn convert_parse_error<I>(
    spans: &SpanStorage<Span>,
    parent: SpanId,
    error: ParseError<I, ContextError>,
) -> (Label<SpanId>, Option<String>) {
    let offset = error.offset();
    let error = error.into_inner();

    let span = spans.insert(Span {
        range: TextRange::empty(TextSize::new(offset as u32)),
        pointer: None,

        parent_id: Some(parent),
    });

    // adapted from the `Display` for `ContextError`.
    let expression = error.context().find_map(|context| match context {
        StrContext::Label(label) => Some(*label),
        _ => None,
    });

    let message = expression.map_or_else(
        || Cow::Borrowed("Syntax error"),
        |expr| Cow::Owned(format!("Invalid {expr}")),
    );

    let label = Label::new(span, message);

    let expected: Vec<_> = error
        .context()
        .filter_map(|context| match context {
            StrContext::Expected(expected) => Some(expected),
            _ => None,
        })
        .collect();

    let expected = if expected.is_empty() {
        None
    } else {
        let mut buffer = String::new();

        if expected.len() == 1 {
            let _ = write!(buffer, "Expected {}", expected[0]);
        } else {
            buffer.push_str("Expected one of: ");

            for (index, value) in expected.iter().enumerate() {
                match index {
                    0 => {}
                    i if i == expected.len() - 1 => buffer.push_str(" or "),
                    _ => buffer.push_str(", "),
                }

                let _ = write!(buffer, "{value}");
            }
        }

        Some(buffer)
    };

    (label, expected)
}

const SYNTAX_ERROR_NOTE: &str =
    "Check for missing delimiters, incorrect operators, or typos in identifiers.";

pub(crate) fn invalid_expr<I>(
    spans: &SpanStorage<Span>,
    parent: SpanId,
    error: ParseError<I, ContextError>,
) -> StringDiagnostic {
    let mut diagnostic =
        Diagnostic::new(StringDiagnosticCategory::InvalidExpression, Severity::Error);

    let (label, expected) = convert_parse_error(spans, parent, error);

    diagnostic.labels.push(label);

    if let Some(expected) = expected {
        diagnostic.add_help(Help::new(expected));
        diagnostic.add_note(Note::new(SYNTAX_ERROR_NOTE));
    }

    diagnostic
}
