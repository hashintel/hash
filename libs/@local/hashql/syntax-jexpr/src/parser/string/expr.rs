use hashql_ast::node::{
    expr::{
        Expr, ExprKind, FieldExpr, IndexExpr, LiteralExpr,
        literal::{IntegerLiteral, LiteralKind},
    },
    id::NodeId,
};
use hashql_core::symbol::{Ident, IdentKind};
use winnow::{
    ModalResult, Parser as _,
    ascii::{digit1, multispace0},
    combinator::{
        alt, cut_err, delimited, dispatch, eof, fail, peek, preceded, repeat, terminated,
    },
    error::{AddContext, ParserError, StrContext, StrContextValue},
    token::any,
};

use super::{combinator::ws, context::Input, ident::parse_ident, path::parse_path};

#[derive(Debug)]
enum Access<'heap> {
    Index(LiteralExpr<'heap>),
    Field(Ident<'heap>),
}

// super limited set of expressions that are supported in strings for convenience
fn parse_field_access<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Access<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    preceded(
        ws("."),
        alt((
            parse_ident.map(Access::Field),
            digit1.with_span().map(|(digit, range)| {
                Access::Field(Ident {
                    span: context.span(range),

                    value: context.heap.intern_symbol(digit),
                    kind: IdentKind::Lexical, // Do we need to specify a different kind here?
                })
            }),
        )),
    )
    .parse_next(input)
}

fn parse_index_access<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Access<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    // super limited version that only allows literal access instead of arbitrary expressions
    delimited(
        ws("["),
        digit1.with_span().map(|(digit, range)| {
            let span = context.span(range);

            Access::Index(LiteralExpr {
                id: NodeId::PLACEHOLDER,
                span,
                kind: LiteralKind::Integer(IntegerLiteral {
                    value: context.heap.intern_symbol(digit),
                }),
                r#type: None,
            })
        }),
        ws(cut_err("]").context(StrContext::Expected(StrContextValue::CharLiteral(']')))),
    )
    .context(StrContext::Label("index"))
    .parse_next(input)
}

pub(crate) fn parse_expr_path<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Expr<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    (
        parse_path.with_span(),
        repeat(
            0..,
            preceded(
                multispace0,
                dispatch! {peek(any);
                    '[' => parse_index_access,
                    '.' => parse_field_access,
                    _ => fail
                        .context(StrContext::Expected(StrContextValue::CharLiteral('[')))
                        .context(StrContext::Expected(StrContextValue::CharLiteral('.')))
                }
                .with_span(),
            ),
        ),
    )
        .map(|((path, path_span), access): (_, Vec<_>)| {
            let mut range = path_span;

            let mut expr = Expr {
                id: NodeId::PLACEHOLDER,
                span: path.span,
                kind: ExprKind::Path(path),
            };

            for (access, span) in access {
                range.end = span.end;

                let span = context.span(range.clone());

                let kind = match access {
                    Access::Index(literal) => ExprKind::Index(IndexExpr {
                        id: NodeId::PLACEHOLDER,
                        span,
                        value: context.heap.boxed(expr),
                        index: context.heap.boxed(Expr {
                            id: NodeId::PLACEHOLDER,
                            span: literal.span,
                            kind: ExprKind::Literal(literal),
                        }),
                    }),
                    Access::Field(ident) => ExprKind::Field(FieldExpr {
                        id: NodeId::PLACEHOLDER,
                        span,
                        value: context.heap.boxed(expr),
                        field: ident,
                    }),
                };

                expr = Expr {
                    id: NodeId::PLACEHOLDER,
                    span,
                    kind,
                };
            }

            expr
        })
        .parse_next(input)
}

// super limited set of expressions that are supported in strings for convenience
pub(crate) fn parse_expr<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Expr<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    alt((
        terminated('_', eof).span().map(|span| Expr {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            kind: ExprKind::Underscore,
        }),
        parse_expr_path,
    ))
    .context(StrContext::Label("expression"))
    .parse_next(input)
}

#[cfg(test)]
mod tests {
    use super::{parse_expr, parse_field_access, parse_index_access};
    use crate::parser::string::test::{bind_parser, test_cases};

    // Bind our parsers to create testing functions
    bind_parser!(SyntaxDump; fn parse_expr_test(parse_expr));
    bind_parser!(Debug; fn parse_field_access_test(parse_field_access));
    bind_parser!(Debug; fn parse_index_access_test(parse_index_access));

    // Tests for field access
    test_cases!(parse_field_access_test;
        simple_field(".field") => "Simple field access",
        numeric_field(".123") => "Numeric field name",
        whitespace_before(".  field") => "Field access with whitespace before name",

        // Error cases
        missing_field_name(".") => "Missing field name",
        invalid_field_name(".@field") => "Invalid field name",
    );

    // Tests for index access
    test_cases!(parse_index_access_test;
        simple_index("[0]") => "Simple index access",
        multi_digit_index("[123]") => "Multi-digit index",
        whitespace_in_brackets("[ 42 ]") => "Whitespace inside brackets",

        // Error cases
        empty_brackets("[]") => "Empty brackets",
        unclosed_bracket("[42") => "Unclosed bracket",
        non_numeric_index("[abc]") => "Non-numeric index",
    );

    // Tests for full expressions
    test_cases!(parse_expr_test;
        // Simple paths
        simple_identifier("foo") => "Simple identifier",
        qualified_path("std::collections::HashMap") => "Qualified path",

        // Field access
        single_field_access("foo.bar") => "Single field access",
        chained_field_access("foo.bar.baz") => "Chained field access",
        numeric_field_access("foo.123") => "Numeric field access",

        // Index access
        single_index_access("foo[0]") => "Single index access",
        chained_index_access("foo[0][1]") => "Chained index access",

        // Mixed access
        mixed_access_1("foo.bar[0]") => "Field then index access",
        mixed_access_2("foo[0].bar") => "Index then field access",
        complex_access("foo.bar[0].baz[42].qux") => "Complex mixed access chain",

        // Path with generics
        generic_path("Vec<Int>") => "Path with generic",
        generic_with_access("Vec<Int>[0]") => "Generic path with index access",
        generic_with_field("Result<T, E>.value") => "Generic path with field access",

        // Whitespace handling
        whitespace_between_access("foo . bar [ 0 ]") => "Whitespace between access operations",
        whitespace_in_path("std :: collections :: HashMap") => "Whitespace in path",

        // Rooted paths
        rooted_path("::foo") => "Rooted path",
        rooted_with_access("::foo.bar[0]") => "Rooted path with access",

        // Underscore paths
        underscore("_") => "Underscore path",
        underscore_identifier("_name") => "Underscore identifier",
    );

    // Error cases for full expressions
    test_cases!(parse_expr_test;
        invalid_start(".foo") => "Starting with field access",
        invalid_start_2("[0]") => "Starting with index access",
        empty_expression("") => "Empty expression",
        unclosed_access_chain("foo.bar[0") => "Unclosed access chain",
        invalid_character("foo$bar") => "Invalid character in expression",
    );
}
