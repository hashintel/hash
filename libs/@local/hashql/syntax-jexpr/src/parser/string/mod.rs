mod combinator;
mod context;
pub(crate) mod error;
mod expr;
mod generic;
mod ident;
mod path;
#[cfg(test)]
pub(crate) mod test;
mod r#type;

use core::fmt::Debug;

use hashql_ast::node::{expr::Expr, r#type::Type};
use hashql_core::{span::SpanId, symbol::Ident};
use winnow::{
    LocatingSlice, ModalParser, Stateful,
    error::{ContextError, ParseError},
    stream::{Offset, Stream, StreamIsPartial},
};

use self::{
    context::Context,
    error::{StringDiagnostic, invalid_expr},
    expr::parse_expr,
    ident::{parse_ident, parse_ident_labelled_argument},
    r#type::parse_type,
};
use super::state::ParserState;
use crate::{
    lexer::{token::Token, token_kind::TokenKind},
    span::Span,
};

type InputStream<'heap, 'span, I> = Stateful<LocatingSlice<I>, Context<'heap, 'span>>;

fn parse_from_string<'heap, 'span, I, O>(
    mut parser: impl ModalParser<InputStream<'heap, 'span, I>, O, ContextError>,
    state: &'span ParserState<'heap, '_>,
    parent: SpanId,
    input: I,
) -> Result<O, ParseError<InputStream<'heap, 'span, I>, ContextError>>
where
    I: Stream + StreamIsPartial + Offset + Clone,
{
    let context = Context {
        heap: state.heap(),
        spans: state.spans(),
        parent,
    };

    parser.parse(Stateful {
        input: LocatingSlice::new(input),
        state: context,
    })
}

pub(crate) fn parse_expr_from_string<'heap>(
    state: &ParserState<'heap, '_>,
    parent: SpanId,
    value: &str,
) -> Result<Expr<'heap>, StringDiagnostic> {
    let expr = parse_from_string(parse_expr, state, parent, value);

    match expr {
        Ok(expr) => Ok(expr),
        Err(error) => Err(invalid_expr(state.spans(), parent, error)),
    }
}

#[expect(
    clippy::panic_in_result_fn,
    reason = "If this happened, the contract with the function has been violated, therefore is \
              fatal"
)]
pub(crate) fn parse_string<'heap, 'source>(
    state: &ParserState<'heap, 'source>,
    token: Token<'source>,
) -> Result<Expr<'heap>, StringDiagnostic> {
    let TokenKind::String(value) = token.kind else {
        panic!("Expected string token")
    };

    let id = state.insert_span(Span {
        range: token.span,
        pointer: Some(state.current_pointer()),
        parent_id: None,
    });

    parse_expr_from_string(state, id, &value)
}

pub(crate) fn parse_type_from_string<'heap>(
    state: &ParserState<'heap, '_>,
    parent: SpanId,
    value: &str,
) -> Result<Type<'heap>, StringDiagnostic> {
    let expr = parse_from_string(parse_type, state, parent, value);

    match expr {
        Ok(expr) => Ok(expr),
        Err(error) => Err(invalid_expr(state.spans(), parent, error)),
    }
}

#[expect(
    clippy::panic_in_result_fn,
    reason = "If this happened, the contract with the function has been violated, therefore is \
              fatal"
)]
pub(crate) fn parse_type_from_token<'heap, 'source>(
    state: &ParserState<'heap, 'source>,
    token: Token<'source>,
) -> Result<Type<'heap>, StringDiagnostic> {
    let TokenKind::String(value) = token.kind else {
        panic!("Expected string token")
    };

    let id = state.insert_span(Span {
        range: token.span,
        pointer: Some(state.current_pointer()),
        parent_id: None,
    });

    parse_type_from_string(state, id, &value)
}

pub(crate) fn parse_ident_from_string<'heap>(
    state: &ParserState<'heap, '_>,
    parent: SpanId,
    value: &str,
) -> Result<Ident<'heap>, ParseError<impl Debug, ContextError>> {
    parse_from_string(parse_ident, state, parent, value)
}

pub(crate) fn parse_ident_labelled_argument_from_string<'heap>(
    state: &ParserState<'heap, '_>,
    parent: SpanId,
    value: &str,
) -> Result<Ident<'heap>, ParseError<impl Debug, ContextError>> {
    parse_from_string(parse_ident_labelled_argument, state, parent, value)
}
