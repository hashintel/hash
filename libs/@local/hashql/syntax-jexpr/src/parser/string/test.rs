use core::fmt::Debug;

use hashql_ast::heap::Heap;
use hashql_core::span::{TextRange, storage::SpanStorage};
use text_size::TextSize;
use winnow::{LocatingSlice, Parser as _, Stateful, error::ContextError};

use crate::{parser::string::context::Context, span::Span};

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
#[repr(u32)]
pub(crate) enum ResultKind {
    Ok,
    Err,
}

impl ResultKind {
    fn into_content(self) -> insta::internals::Content {
        insta::internals::Content::UnitVariant(
            "ResultKind",
            self as u32,
            match self {
                Self::Ok => "Ok",
                Self::Err => "Err",
            },
        )
    }
}

#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub(crate) struct Info {
    kind: ResultKind,
}

impl Info {
    pub(crate) fn into_content(self) -> insta::internals::Content {
        insta::internals::Content::Struct("Info", vec![("kind", self.kind.into_content())])
    }
}

macro format_result {
    (Debug; $ident:ident) => {
        format!("{:#?}", $ident)
    },
    (SyntaxDump; $ident:ident) => {
        hashql_ast::format::SyntaxDump::syntax_dump_to_string(&$ident)
    }
}

pub(crate) macro bind_parser($format:ident; fn $name:ident($parser:ident)) {
    // this could also be a function, but then you run into issues with lifetimes, so this is easier
    fn $name(source: &str) -> (String, Info) {
        let heap = Heap::new();
        let spans = SpanStorage::new();
        let parent = spans.insert(Span {
            range: TextRange::up_to(TextSize::of(source)),
            pointer: None,
            parent_id: None,
        });

        let context = Context {
            heap: &heap,
            spans: &spans,
            parent,
        };

        let input = Stateful {
            input: LocatingSlice::new(source),
            state: context,
        };

        let result = $parser::<ContextError>.parse(input);
        match result {
            Ok(ident) => (
                format_result!($format; ident),
                Info {
                    kind: ResultKind::Ok,
                },
            ),
            Err(error) => (
                error.to_string(),
                Info {
                    kind: ResultKind::Err,
                },
            ),
        }
    }
}

pub(crate) macro assert_parse($parser:ident, $source:expr, $description:literal) {{
        let (result, info) = $parser($source);

        ::insta::with_settings!({
            description => $description,
            raw_info => &info.into_content(),
        }, {
            ::insta::assert_snapshot!(insta::_macro_support::AutoName, result, $source);
        })
    }}

pub(crate) macro test_cases($parser:ident;
        $(
            $name:ident($source:expr) => $description:expr,
        )*
    ) {
        $(
            #[test]
            fn $name() {
                assert_parse!($parser, $source, $description);
            }
        )*
    }
