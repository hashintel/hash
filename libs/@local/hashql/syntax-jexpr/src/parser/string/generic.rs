use hashql_ast::node::{
    generic::{GenericArgument, GenericParam, Generics},
    id::NodeId,
};
use winnow::{
    ModalResult, Parser as _,
    combinator::{cut_err, delimited, opt, preceded},
    error::{AddContext, ParserError, StrContext, StrContextValue},
};

use super::{
    combinator::{separated_boxed1, ws},
    context::Input,
    ident::parse_ident,
    r#type::parse_type,
};

pub(crate) fn parse_generic_argument<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<GenericArgument<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    parse_type
        .with_span()
        .map(|(r#type, span)| GenericArgument {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            r#type: context.heap.boxed(r#type),
        })
        .context(StrContext::Label("generic argument"))
        .parse_next(input)
}

fn parse_generic_param<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<GenericParam<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    (parse_ident, opt(preceded(ws(":"), parse_type)))
        .with_span()
        .map(|((name, bound), span)| GenericParam {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            name,
            bound: bound.map(|bound| context.heap.boxed(bound)),
        })
        .context(StrContext::Label("generic parameter"))
        .parse_next(input)
}

#[expect(clippy::allow_attributes, reason = "except doesn't work here")]
#[allow(
    dead_code,
    reason = "This function will be used in the future for parsing closure signatures"
)]
pub(crate) fn parse_generics<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Generics<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    delimited(
        ws("<"),
        separated_boxed1(context.heap, parse_generic_param, ws(",")),
        ws(cut_err(">").context(StrContext::Expected(StrContextValue::CharLiteral('>')))),
    )
    .with_span()
    .map(|(params, span)| Generics {
        id: NodeId::PLACEHOLDER,
        span: context.span(span),
        params,
    })
    .context(StrContext::Label("generics"))
    .parse_next(input)
}

#[cfg(test)]
mod tests {
    use super::{parse_generic_argument, parse_generics};
    use crate::parser::string::test::{bind_parser, test_cases};

    // Bind our parsers to create testing functions
    bind_parser!(SyntaxDump; fn parse_generic_argument_test(parse_generic_argument));
    bind_parser!(SyntaxDump; fn parse_generics_test(parse_generics));

    // Tests for generic arguments
    test_cases!(parse_generic_argument_test;
        simple_type("Int") => "Simple type as generic argument",
        complex_type("Map<K, V>") => "Generic type as generic argument",
        nested_type("Option<Result<T, E>>") => "Nested generic types",
        union_type("Int | String") => "Union type as generic argument",
        intersection_type("Serializable & Equatable") => "Intersection type as generic argument",
        tuple_type("(Int, String)") => "Tuple type as generic argument",
        struct_type("(key: K, value: V)") => "Struct type as generic argument",
        infer_type("_") => "Infer type as generic argument",
        whitespace_handling_argument(" Int ") => "Generic argument with whitespace",
    );

    // Tests for generics (lists of generic parameters)
    test_cases!(parse_generics_test;
        single_param("<T>") => "Single generic parameter without bound",
        multiple_params("<T, U, V>") => "Multiple generic parameters",

        // With type bounds
        param_with_bound("<T: Object>") => "Generic parameter with type bound",
        mixed_bounds("<T, U: Comparable, V>") => "Mix of bounded and unbounded parameters",
        complex_bound("<K: Map<T, V>>") => "Complex type bound",
        union_bound("<T: A | B>") => "Union type as bound",
        intersection_bound("<T: Serializable & Comparable>") => "Intersection type as bound",

        // Whitespace variations
        whitespace_handling("< T , U >") => "Generics with whitespace",
        bound_whitespace("< T : Comparable >") => "Bound with whitespace",

        // Error cases
        unclosed_generics("<T, U") => "Unclosed generics",
        empty_generics("<>") => "Empty generics",
        missing_comma("<T U>") => "Missing comma between parameters",
        invalid_bound("<T: >") => "Invalid bound specification",
        invalid_parameter("<123>") => "Invalid parameter name",
    );
}
