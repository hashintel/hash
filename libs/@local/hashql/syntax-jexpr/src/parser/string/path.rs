use hashql_ast::node::{
    id::NodeId,
    path::{Path, PathSegment},
};
use winnow::{
    ModalResult, Parser as _,
    combinator::{cut_err, delimited, opt},
    error::{AddContext, ParserError, StrContext, StrContextValue},
};

use super::{
    combinator::{separated_boxed1, ws},
    context::Input,
    generic::parse_generic_argument,
    ident::parse_ident,
};

fn parse_path_segment<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<PathSegment<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let arguments = opt(delimited(
        ws("<"),
        separated_boxed1(context.heap, parse_generic_argument, ws(",")),
        ws(cut_err(">").context(StrContext::Expected(StrContextValue::CharLiteral('>')))),
    ));

    (parse_ident, arguments)
        .with_span()
        .map(|((ident, arguments), span)| PathSegment {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            name: ident,
            arguments: arguments.unwrap_or_else(|| context.heap.empty_slice()),
        })
        .parse_next(input)
}

pub(crate) fn parse_path<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Path<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let root = opt(ws("::")).map(|value| value.is_some());
    let segments = separated_boxed1(context.heap, parse_path_segment, ws("::"));

    (root, segments)
        .with_span()
        .map(|((rooted, segments), span)| Path {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            rooted,
            segments,
        })
        .context(StrContext::Label("path"))
        .parse_next(input)
}

#[cfg(test)]
mod tests {
    #![expect(clippy::non_ascii_literal)]
    use super::parse_path;
    use crate::parser::string::test::{bind_parser, test_cases};

    // Bind our parser to create a testing function
    bind_parser!(SyntaxDump; fn parse_path_test(parse_path));

    // Basic path structure tests
    test_cases!(parse_path_test;
        simple_identifier("foo") => "Simple identifier path",
        two_segment_path("foo::bar") => "Two-segment path",
        multi_segment_path("foo::bar::baz") => "Multi-segment path",
        rooted_path("::foo") => "Rooted path",
        rooted_multi_segment("::foo::bar") => "Rooted multi-segment path",
    );

    // Generic arguments
    test_cases!(parse_path_test;
        single_generic("Vec<T>") => "Single generic argument",
        multiple_generics("Map<K, V>") => "Multiple generic arguments",
        path_with_generics("std::collections::HashMap<K, V>") => "Path segments with generics",
        nested_generics("Result<Option<T>, E>") => "Nested generic arguments",
        complex_generics("HashMap<String, Vec<Option<T>>>") => "Complex nested generics",
    );

    // Whitespace variations
    test_cases!(parse_path_test;
        path_whitespace("foo :: bar") => "Whitespace around path separator",
        generics_whitespace("Vec< T , U >") => "Whitespace in generic arguments",
        mixed_whitespace("std :: collections :: HashMap< K , V >") => "Mixed whitespace patterns",
    );

    // Edge cases
    test_cases!(parse_path_test;
        unicode_identifier("标识符") => "Unicode identifier",
        mixed_unicode("std::标识符::Vec<T>") => "Mixed Unicode and ASCII in path",
        path_trailing_separator("foo::") => "Path with trailing separator",
        empty_root("::") => "Empty rooted path",
    );

    // Error handling
    test_cases!(parse_path_test;
        unclosed_generic("Vec<T") => "Unclosed generic argument",
        empty_generic_args("Vec<>") => "Empty generic arguments",
        missing_separator("Vec<T U>") => "Missing separator in generic arguments",
        invalid_separator("foo;bar") => "Invalid path separator",
        double_separator("foo::::bar") => "Double path separator",
        empty_path("") => "Empty path",
    );
}
