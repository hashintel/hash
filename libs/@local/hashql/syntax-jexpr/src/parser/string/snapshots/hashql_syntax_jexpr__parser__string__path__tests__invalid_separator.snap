---
source: libs/@local/hashql/syntax-jexpr/src/parser/string/path.rs
description: Invalid path separator
expression: foo;bar
info:
  kind: Err
---
foo;bar
   ^
